### PR TITLE
Move GIT credentials to Organization level

### PR DESCRIPTION
**Pull request by nulls**
### What's done:
- replaced project_id to organization_id in table git

### CODE REVIEW

File: save-backend/src/test/kotlin/com/saveourtool/save/backend/controller/ProjectControllerTest.kt [hunk 1]
```diff
@@ -186,14 +186,12 @@ class ProjectControllerTest {
             details = AuthenticationDetails(id = 2)
         }
 
-        val gitDto = GitDto("qweqwe")
```
Comment (nulls) [inline save-backend/src/test/kotlin/com/saveourtool/save/backend/controller/ProjectControllerTest.kt:189]: need to add test for `OrganizationService`

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt [hunk 2]
```diff
@@ -93,6 +95,27 @@ private fun organizationSettingsMenu() = FC<OrganizationSettingsMenuProps> { pro
...
+                className = ClassName("text-xs text-center font-weight-bold text-primary text-uppercase mb-3")
+                +"Git credentials"
+            }
+            organizationGitCredentialsManageCard {
+                selfUserInfo = props.currentUserInfo
+                organizationName = props.organizationName
+                this.wasConfirmationModalShown = wasConfirmationModalShown
+                updateErrorMessage = props.updateErrorMessage
+                getUserGroups = { it.organizations }
+                showGlobalRoleWarning = {
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt:110]: move to common function

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt [hunk 3]
```diff
@@ -457,6 +454,16 @@ class OrganizationView : AbstractView<OrganizationProps, OrganizationViewState>(
         }
     }
 
+    private fun ChildrenBuilder.renderTests() {
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt:457]: just a stub

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt [hunk 4]
```diff
@@ -562,6 +569,16 @@ class OrganizationView : AbstractView<OrganizationProps, OrganizationViewState>(
         it.decodeFromJsonString<ImageInfo>()
     }
 
+    private suspend fun getGits(): List<GitDto> = get(
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt:572]: it's not used

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt [hunk 5]
```diff
@@ -641,7 +658,7 @@ class OrganizationView : AbstractView<OrganizationProps, OrganizationViewState>(
                                             }
                                         }
                                     }
-                                    +projectMenu.name
+                                    +projectMenu.getTitle()
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt:661]: added to be able to have title with whitespaces in name

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt [hunk 6]
```diff
@@ -331,37 +338,18 @@ class ProjectView : AbstractView<ProjectExecutionRouteProps, ProjectViewState>(f
...
-            val gitDtoInit: GitDto = post(
-                "$apiUrl/projects/git",
-                headers,
-                jsonProject,
-                loadingHandler = ::noopLoadingHandler,
-            ).decodeFromJsonString()
-            gitDto = gitDtoInit
             val currentUserRole: Role = get(
-                "$apiUrl/projects/${state.project.organization.name}/${state.project.name}/users/roles",
+                "$apiUrl/projects/${project.organization.name}/${project.name}/users/roles",
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt:346]: project in stub is a stub

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt [hunk 7]
```diff
@@ -338,6 +334,27 @@ class ExecutionController(private val executionService: ExecutionService,
         }
         .orEmpty()
 
+    @Suppress("UnsafeCallOnNullableType")
```
Comment (nulls) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt:337]: there is no `UnsafeCallOnNullableType`, need to remove

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt [hunk 8]
```diff
@@ -83,16 +85,25 @@ private fun organizationSettingsMenu() = FC<OrganizationSettingsMenuProps> { pro
...
+            div {
+                className = ClassName("text-xs text-center font-weight-bold text-primary text-uppercase mb-3")
+                +"Git credentials"
+            }
+            organizationGitCredentialsManageCard {
+                selfUserInfo = props.currentUserInfo
+                organizationName = props.organizationName
+                this.wasConfirmationModalShown = wasConfirmationModalShown
+                updateErrorMessage = props.updateErrorMessage
+                getUserGroups = { it.organizations }
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt:103]: it's not clear where it's used...

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/CreationView.kt [hunk 9]
```diff
@@ -379,7 +373,7 @@ class CreationView : AbstractView<Props, ProjectSaveViewState>(true) {
                                                 +"Provide Credentials if your repo with Test Suites is private:"
                                             }
                                         }
-
+                                        // FIXME: need to move to organization window
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/CreationView.kt:376]: already moved, need to remove

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt [hunk 10]
```diff
@@ -217,13 +222,14 @@ external interface ProjectViewState : State {
...
     private var standardTestSuites: List<TestSuiteDto> = emptyList()
     private val selectedStandardSuites: MutableList<String> = mutableListOf()
-    private var gitDto: GitDto? = null
     private val date = LocalDateTime(1970, Month.JANUARY, 1, 0, 0, 1)
     private val testResourcesSelection = testResourcesSelection(
         updateGitUrlFromInputField = { event ->
-            event.preventDefault()
             setState {
-                gitUrlFromInputField = event.target.value
+                availableGitCredentials.find { it.url == event.target.value }
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt:229]: all callback can be replaced to string callbacks

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt [hunk 11]
```diff
@@ -981,6 +949,17 @@ class ProjectView : AbstractView<ProjectExecutionRouteProps, ProjectViewState>(f
             it.decodeFromJsonString<List<FileInfo>>()
         }
 
+    private suspend fun getGitCredentials(
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt:952]: we call this method in two places -- can be combined

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+            .unsafeMap {
+                it.decodeFromJsonString<List<GitDto>>()
+            }.let {
+                setGitCredentials(it)
+            }
+    }
+    return Triple(gitCredentials, setGitCredentials, fetchGitCredentialsRequest)
+}
+
+private fun prepareUpsertGitCredential(
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt:222]: need to check that we add gitUrl which is not added yet

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/CreationView.kt (hunk 9 above)
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/CreationView.kt:376]: need to move connectivity validation too

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt (hunk 7 above)
Comment (nulls) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt:337]: removed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt (hunk 8 above)
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt:103]: removed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt (hunk 10 above)
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt:229]: changed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt (hunk 12 above)
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt:222]: agreed to leave current logic -- if we add a new git credentials with existed url -- it updates existed value

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/CreationView.kt (hunk 9 above)
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/CreationView.kt:376]: agreed to leave it as is and move it in next phase

File: save-api/src/main/kotlin/com/saveourtool/save/api/SaveCloudClient.kt [hunk 13]
```diff
@@ -126,6 +124,7 @@ class SaveCloudClient(
         return ExecutionRequest(
             project = project,
             gitDto = gitDto,
+            branchOrCommit = evaluatedToolProperties.branch ?: evaluatedToolProperties.commitHash,
```
Comment (nulls) [inline save-api/src/main/kotlin/com/saveourtool/save/api/SaveCloudClient.kt:127]: repeat logic from `DownloadProjectController`

File: db/v-2/insert-test-data/git-insert.xml [hunk 14]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
+<changeSet id="organization_id-old-project" author="nulls" context="dev">
```
Comment (nulls) [inline db/v-2/insert-test-data/git-insert.xml:7]: old projects in test data don't contain `organization_id`

File: db/v-2/insert-test-data/test-suite-insert.xml [hunk 15]
```diff
@@ -98,5 +98,18 @@
...
+
+        <insert dbms="mysql" tableName="test_suite">
+            <column name="id" value="11" />
+            <column name="name" value="test" />
+            <column name="type" value="STANDARD" />
+            <column name="project_id" value="2" />
+            <column name="date_added" value="2021-01-01 00:00:00"/>
+            <column name="test_root_path" value=""/>
+            <column name="description" value="description"/>
+            <column name="test_suite_repo_url" value="https://github.com/saveourtool/save-cloud"/>
```
Comment (nulls) [inline db/v-2/insert-test-data/test-suite-insert.xml:110]: need to have a valid `test_suite_repo_url`

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+import react.dom.html.ButtonType
+import react.dom.html.ReactHTML.button
+import react.dom.html.ReactHTML.div
+import react.dom.html.ReactHTML.hr
+import react.useState
+
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+typealias RequestWithDependency<R> = Triple<R, StateSetter<R>, () -> Unit>
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt:31]: common approach -- can be extracted

File: save-backend/src/test/kotlin/com/saveourtool/save/backend/controller/ProjectControllerTest.kt (hunk 1 above)
Comment (nulls) [reply save-backend/src/test/kotlin/com/saveourtool/save/backend/controller/ProjectControllerTest.kt:189]: added

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+        deleteGitCredentialRequest()
+        setConfirmDeleteGitCredentialWindowOpened(false)
+    }
+
+    val canModify = selfRole == Role.SUPER_ADMIN || selfRole == Role.ADMIN
+    div {
+        className = ClassName("card card-body mt-0 pt-0 pr-0 pl-0")
+        for (gitCredential in gitCredentials) {
+            val url = gitCredential.url
+            val index = gitCredentials.indexOf(gitCredential)
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt:125]: `gitCredentials.forEachIndexed`?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/ReactUtils.kt [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+package com.saveourtool.save.frontend.utils
+
+import react.useState
+
+/**
+ * Runs the provided [action] only once of first render
+ *
+ * @param action
+ */
+fun runOnlyOnFirstRender(action: () -> Unit) {
```
[the remark this answers is not in the document]
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/ReactUtils.kt:14]: sure

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt (hunk 17 above)
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/ManageGitCredentialsCard.kt:125]: done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/ReactUtils.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/ReactUtils.kt:14]: added `ChildrenBuilder` as receiver

[the remark this answers is not in the document]
Comment (petertrr) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/ReactUtils.kt]: `ChildrenBuilder` can be a receiver in both function and class components, however inside the function you are using hooks that can only be called from function components.

**Review by petertrr — APPROVED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/ReactUtils.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/ReactUtils.kt]: discussed offline, that we cannot check that function is called from `FC`

**Pull request merged**